Adding nulling and explicit columns filter to outputs

amitschang opened the pull request:
In cases where an entire column has no non-null values, the mapping will be missing from the json entirely and as such column will be missing. Since the database still expects values we need to provide these null-filled. Being more explicit here, while a little clunky, has the advantage that we can fail fast when required columns are missing.  The view of required and nullable from the DB perspective is not necessarily the same as for the json data, optional here means the entire dataset could be missing the column.

Also adding deduplication to the dyads table, which has been seen to have redundant entries for model for example.

amitschang commented on susdingest/jsonloader.py:
```diff
@@ -56,27 +56,37 @@ def normalize_col(self, column, explode=True):
+        publications = self.with_nullable_as_required(self.data.copy(), optional_cols + ['expressions'], required_cols)
+        publications['journal_issn_isbn'] = publications['journal_issn_isbn'] \
+            .apply(lambda x: '|'.join(str(i) for i in x) if x == x else x)
+        publications['tested_expressions'] = publications['expressions'] \
```
it is possible, it is added above as an optional col.

glemson replied:
ah yes, missed that statement. but should the column name not be 'tested_expressions'

amitschang commented on susdingest/jsonloader.py:
```diff
@@ -56,27 +56,37 @@ def normalize_col(self, column, explode=True):
+                            .rename(columns=lambda x: f'journal_{x}')
+            publications = publications.join(citescore)
+        return self.with_nullable_as_required(
+            publications, required_cols, optional_cols + ['journal_citescore_year', 'journal_citescore_value']) \
```
quite right. I suppose I need to add a test to excersize all optional columns - would have caught this. Thanks!

amitschang commented:
I've updated and added tests for the optional/required columns. Shouldn't have skimped in the first place since it brought out an issue in null checks, which don't work the same between pd.NA and np.nan null types, so added a function to coalesce those checks.

glemson reviewed (approved):
go for it
